File: fatty/journal.py
from __future__ import annotations

import json
import os
import tempfile
import threading
import uuid
from collections.abc import Callable
from dataclasses import asdict, dataclass
from datetime import datetime
from pathlib import Path

from fatty.store import APP_DIR

JOURNAL_PATH = APP_DIR / "journal.jsonl"
MAX_ENTRIES = 5000
_TRIM_SLACK = 200
_COMMAND_MAX = 16_384
_ERROR_MAX = 4_096

Listener = Callable[[], None]


@dataclass
class JournalEntry:
    id: str
    started_at: str
    finished_at: str
    duration_sec: float
    server_id: str
    server_name: str
    host: str
    port: int
    username: str
    command_id: str = ""
    title: str = ""
    command: str = ""
    cwd: str = ""
    login_shell: bool = True
    timeout_sec: int = 180
    exit_code: int | None = None
    status: str = "error"
    kind: str = "command"
    error: str = ""

    def target(self) -> str:
        return f"{self.username}@{self.host}:{self.port}"

    def started_display(self) -> str:
        return _format_when(self.started_at)

    def duration_display(self) -> str:
        return format_duration(self.duration_sec)

    def status_display(self) -> str:
        if self.exit_code is not None and self.status in {"ok", "failed"}:
            return str(self.exit_code)
        return STATUS_LABELS.get(self.status, self.status)

    def command_preview(self, limit: int = 80) -> str:
        text = " ".join((self.command or "").split())
        if len(text) > limit:
            return text[: limit - 1] + "…"
        return text

    def as_text(self) -> str:
        lines = [
            f"Время: {self.started_display()}",
            f"Длительность: {self.duration_display()}",
            f"VPS: {self.server_name}  ({self.target()})",
            f"Тип: {KIND_LABELS.get(self.kind, self.kind)}",
            f"Название: {self.title or '—'}",
        ]
        if self.cwd:
            lines.append(f"Каталог: {self.cwd}")
        result = self.status_display()
        if self.exit_code is not None and self.status not in {"ok", "failed"}:
            result = f"{STATUS_LABELS.get(self.status, self.status)} (код {self.exit_code})"
        lines.append(f"Результат: {result}")
        if self.error:
            lines.append(f"Ошибка: {self.error}")
        lines.append("Команда:")
        lines.append(self.command or "—")
        return "\n".join(lines)


STATUS_LABELS = {
    "ok": "OK",
    "failed": "ошибка",
    "timeout": "таймаут",
    "cancelled": "прервано",
    "error": "сбой",
}

KIND_LABELS = {
    "command": "сохранённая",
    "quick": "разовая",
    "test": "проверка связи",
}


def now_iso() -> str:
    return datetime.now().astimezone().isoformat(timespec="seconds")


def format_duration(seconds: float) -> str:
    if seconds < 0:
        seconds = 0.0
    if seconds < 60:
        return f"{seconds:.1f} с"
    total = int(round(seconds))
    minutes, sec = divmod(total, 60)
    if minutes < 60:
        return f"{minutes} мин {sec} с"
    hours, minutes = divmod(minutes, 60)
    return f"{hours} ч {minutes} мин"


def status_from_exit(code: int) -> str:
    if code == 0:
        return "ok"
    if code == 124:
        return "timeout"
    if code == 130:
        return "cancelled"
    return "failed"


def _format_when(value: str) -> str:
    raw = (value or "").strip()
    if not raw:
        return "—"
    try:
        dt = datetime.fromisoformat(raw)
        return dt.strftime("%Y-%m-%d %H:%M:%S")
    except ValueError:
        return raw.replace("T", " ")[:19]


def _clip(text: str, limit: int) -> str:
    text = text or ""
    if len(text) <= limit:
        return text
    return text[: limit - 1] + "…"


def _parse_entry(raw: dict) -> JournalEntry | None:
    try:
        port = int(raw.get("port", 22) or 22)
    except (TypeError, ValueError):
        port = 22
    try:
        duration = float(raw.get("duration_sec", 0) or 0)
    except (TypeError, ValueError):
        duration = 0.0
    try:
        timeout = int(raw.get("timeout_sec", 180) or 180)
    except (TypeError, ValueError):
        timeout = 180
    exit_raw = raw.get("exit_code", None)
    exit_code: int | None
    if exit_raw is None or exit_raw == "":
        exit_code = None
    else:
        try:
            exit_code = int(exit_raw)
        except (TypeError, ValueError):
            exit_code = None
    entry_id = str(raw.get("id") or "").strip() or str(uuid.uuid4())
    return JournalEntry(
        id=entry_id,
        started_at=str(raw.get("started_at") or ""),
        finished_at=str(raw.get("finished_at") or ""),
        duration_sec=max(0.0, duration),
        server_id=str(raw.get("server_id") or ""),
        server_name=str(raw.get("server_name") or ""),
        host=str(raw.get("host") or ""),
        port=port,
        username=str(raw.get("username") or ""),
        command_id=str(raw.get("command_id") or ""),
        title=str(raw.get("title") or ""),
        command=str(raw.get("command") or ""),
        cwd=str(raw.get("cwd") or ""),
        login_shell=bool(raw.get("login_shell", True)),
        timeout_sec=timeout,
        exit_code=exit_code,
        status=str(raw.get("status") or "error"),
        kind=str(raw.get("kind") or "command"),
        error=str(raw.get("error") or ""),
    )


def _write_lines(path: Path, lines: list[str]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp_name = tempfile.mkstemp(prefix="journal-", suffix=".jsonl", dir=path.parent)
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as handle:
            if lines:
                handle.write("\n".join(lines))
                handle.write("\n")
        os.replace(tmp_name, path)
    except Exception:
        if os.path.exists(tmp_name):
            os.remove(tmp_name)
        raise


class Journal:
    def __init__(self, path: Path | None = None, *, max_entries: int = MAX_ENTRIES) -> None:
        self.path = path or JOURNAL_PATH
        self.max_entries = max(100, min(50_000, int(max_entries or MAX_ENTRIES)))
        self._lock = threading.Lock()
        self._listeners: list[Listener] = []

    def add_listener(self, callback: Listener) -> None:
        if callback not in self._listeners:
            self._listeners.append(callback)

    def remove_listener(self, callback: Listener) -> None:
        try:
            self._listeners.remove(callback)
        except ValueError:
            pass

    def _notify(self) -> None:
        for callback in list(self._listeners):
            try:
                callback()
            except Exception:
                pass

    def append(self, entry: JournalEntry) -> None:
        entry.command = _clip(entry.command, _COMMAND_MAX)
        entry.error = _clip(entry.error, _ERROR_MAX)
        line = json.dumps(asdict(entry), ensure_ascii=False)
        with self._lock:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            with self.path.open("a", encoding="utf-8") as handle:
                handle.write(line)
                handle.write("\n")
            self._trim_unlocked()
        self._notify()

    def load(self, limit: int = MAX_ENTRIES) -> list[JournalEntry]:
        with self._lock:
            return self._read_unlocked(limit)

    def clear(self) -> None:
        with self._lock:
            _write_lines(self.path, [])
        self._notify()

    def export_text(self, entries: list[JournalEntry] | None = None) -> str:
        if entries is None:
            entries = self.load()
        blocks = [item.as_text() for item in entries]
        return ("\n\n" + ("─" * 48) + "\n\n").join(blocks)

    def _read_unlocked(self, limit: int) -> list[JournalEntry]:
        if not self.path.exists():
            return []
        entries: list[JournalEntry] = []
        try:
            text = self.path.read_text(encoding="utf-8")
        except OSError:
            return []
        for line in text.splitlines():
            raw_line = line.strip()
            if not raw_line:
                continue
            try:
                payload = json.loads(raw_line)
            except json.JSONDecodeError:
                continue
            if not isinstance(payload, dict):
                continue
            parsed = _parse_entry(payload)
            if parsed is not None:
                entries.append(parsed)
        if limit > 0 and len(entries) > limit:
            entries = entries[-limit:]
        entries.reverse()
        return entries

    def _trim_unlocked(self) -> None:
        if not self.path.exists():
            return
        try:
            raw_lines = [line for line in self.path.read_text(encoding="utf-8").splitlines() if line.strip()]
        except OSError:
            return
        if len(raw_lines) <= self.max_entries + _TRIM_SLACK:
            return
        _write_lines(self.path, raw_lines[-self.max_entries :])

File: src/core/journal.cpp
#include "core/journal.hpp"

#include "core/paths.hpp"
#include "core/util.hpp"

#include <algorithm>
#include <cmath>
#include <cctype>
#include <fstream>
#include <chrono>
#include <ctime>
#include <iomanip>
#include <nlohmann/json.hpp>
#include <sstream>

namespace fatty {
using json = nlohmann::json;

namespace {

constexpr int kTrimSlack = 200;
constexpr int kCommandMax = 16384;
constexpr int kErrorMax = 4096;

std::tm local_tm(std::chrono::system_clock::time_point tp) {
  std::time_t t = std::chrono::system_clock::to_time_t(tp);
  std::tm out{};
#ifdef _WIN32
  localtime_s(&out, &t);
#else
  localtime_r(&t, &out);
#endif
  return out;
}

std::optional<std::chrono::system_clock::time_point> parse_iso(const std::string& raw) {
  if (raw.empty()) {
    return std::nullopt;
  }
  std::tm tm{};
  std::istringstream ss(raw);
  ss >> std::get_time(&tm, "%Y-%m-%dT%H:%M:%S");
  if (ss.fail()) {
    return std::nullopt;
  }
  return std::chrono::system_clock::from_time_t(std::mktime(&tm));
}

json entry_to_json(const JournalEntry& e) {
  json j = {
      {"id", e.id},
      {"started_at", e.started_at},
      {"finished_at", e.finished_at},
      {"duration_sec", e.duration_sec},
      {"server_id", e.server_id},
      {"server_name", e.server_name},
      {"host", e.host},
      {"port", e.port},
      {"username", e.username},
      {"command_id", e.command_id},
      {"title", e.title},
      {"command", e.command},
      {"cwd", e.cwd},
      {"login_shell", e.login_shell},
      {"timeout_sec", e.timeout_sec},
      {"status", e.status},
      {"kind", e.kind},
      {"error", e.error},
  };
  if (e.exit_code) {
    j["exit_code"] = *e.exit_code;
  } else {
    j["exit_code"] = nullptr;
  }
  return j;
}

JournalEntry parse_entry(const json& raw) {
  JournalEntry e;
  e.id = raw.value("id", "");
  if (e.id.empty()) {
    e.id = new_uuid();
  }
  e.started_at = raw.value("started_at", "");
  e.finished_at = raw.value("finished_at", "");
  e.duration_sec = std::max(0.0, raw.value("duration_sec", 0.0));
  e.server_id = raw.value("server_id", "");
  e.server_name = raw.value("server_name", "");
  e.host = raw.value("host", "");
  e.port = raw.value("port", 22);
  e.username = raw.value("username", "");
  e.command_id = raw.value("command_id", "");
  e.title = raw.value("title", "");
  e.command = raw.value("command", "");
  e.cwd = raw.value("cwd", "");
  e.login_shell = raw.value("login_shell", true);
  e.timeout_sec = raw.value("timeout_sec", 180);
  if (raw.contains("exit_code") && !raw["exit_code"].is_null()) {
    try {
      e.exit_code = raw["exit_code"].get<int>();
    } catch (...) {
    }
  }
  e.status = raw.value("status", "error");
  e.kind = raw.value("kind", "command");
  e.error = raw.value("error", "");
  return e;
}

}  // namespace

std::string status_label(const std::string& status) {
  if (status == "ok") return "OK";
  if (status == "failed") return "ошибка";
  if (status == "timeout") return "таймаут";
  if (status == "cancelled") return "прервано";
  if (status == "error") return "сбой";
  return status;
}

std::string kind_label(const std::string& kind) {
  if (kind == "command") return "сохранённая";
  if (kind == "quick") return "разовая";
  if (kind == "test") return "проверка связи";
  return kind;
}

std::string now_iso() {
  auto now = std::chrono::system_clock::now();
  auto tm = local_tm(now);
  std::ostringstream ss;
  ss << std::put_time(&tm, "%Y-%m-%dT%H:%M:%S");
  return ss.str();
}

std::string format_duration(double seconds) {
  if (seconds < 0) seconds = 0;
  if (seconds < 60) {
    std::ostringstream ss;
    ss.setf(std::ios::fixed);
    ss.precision(1);
    ss << seconds << " с";
    return ss.str();
  }
  int total = static_cast<int>(std::round(seconds));
  int minutes = total / 60;
  int sec = total % 60;
  if (minutes < 60) {
    return std::to_string(minutes) + " мин " + std::to_string(sec) + " с";
  }
  int hours = minutes / 60;
  minutes %= 60;
  return std::to_string(hours) + " ч " + std::to_string(minutes) + " мин";
}

std::string status_from_exit(int code) {
  if (code == 0) return "ok";
  if (code == 124) return "timeout";
  if (code == 130) return "cancelled";
  return "failed";
}

std::string JournalEntry::target() const {
  return username + "@" + host + ":" + std::to_string(port);
}

std::string JournalEntry::started_display() const {
  auto tp = parse_iso(started_at);
  if (!tp) {
    auto t = started_at;
    auto p = t.find('T');
    if (p != std::string::npos) t[p] = ' ';
    return t.empty() ? "—" : t.substr(0, 19);
  }
  auto tm = local_tm(*tp);
  std::ostringstream ss;
  ss << std::put_time(&tm, "%Y-%m-%d %H:%M:%S");
  return ss.str();
}

std::string JournalEntry::duration_display() const {
  return format_duration(duration_sec);
}

std::string JournalEntry::status_display() const {
  if (exit_code && (status == "ok" || status == "failed")) {
    return std::to_string(*exit_code);
  }
  return status_label(status);
}

std::string JournalEntry::command_preview(int limit) const {
  std::string text;
  for (char ch : command) {
    if (std::isspace(static_cast<unsigned char>(ch))) {
      if (!text.empty() && text.back() != ' ') text.push_back(' ');
    } else {
      text.push_back(ch);
    }
  }
  if (static_cast<int>(text.size()) > limit) {
    return text.substr(0, static_cast<std::size_t>(limit - 1)) + "…";
  }
  return text;
}

std::string JournalEntry::compact_when() const {
  auto tp = parse_iso(started_at);
  if (!tp) return "";
  auto tm = local_tm(*tp);
  auto now = local_tm(std::chrono::system_clock::now());
  std::ostringstream ss;
  if (tm.tm_year == now.tm_year && tm.tm_yday == now.tm_yday) {
    ss << std::put_time(&tm, "%H:%M");
  } else if (tm.tm_year == now.tm_year) {
    ss << std::put_time(&tm, "%d.%m %H:%M");
  } else {
    ss << std::put_time(&tm, "%d.%m.%y");
  }
  return ss.str();
}

std::string JournalEntry::last_run_label() const {
  std::string result;
  if (status == "ok") {
    result = "OK";
  } else if (exit_code && status == "failed") {
    result = std::to_string(*exit_code);
  } else {
    result = status_label(status);
  }
  auto when = compact_when();
  if (!when.empty()) {
    return result + " · " + when;
  }
  return result;
}

std::string JournalEntry::as_text() const {
  std::ostringstream ss;
  ss << "Время: " << started_display() << "\n";
  ss << "Длительность: " << duration_display() << "\n";
  ss << "VPS: " << server_name << "  (" << target() << ")\n";
  ss << "Тип: " << kind_label(kind) << "\n";
  ss << "Название: " << (title.empty() ? "—" : title) << "\n";
  if (!cwd.empty()) {
    ss << "Каталог: " << cwd << "\n";
  }
  std::string result = status_display();
  if (exit_code && status != "ok" && status != "failed") {
    result = status_label(status) + " (код " + std::to_string(*exit_code) + ")";
  }
  ss << "Результат: " << result << "\n";
  if (!error.empty()) {
    ss << "Ошибка: " << error << "\n";
  }
  ss << "Команда:\n" << (command.empty() ? "—" : command);
  return ss.str();
}

Journal::Journal(std::filesystem::path path, int max_entries_in)
    : max_entries(std::max(100, std::min(50000, max_entries_in))),
      path_(path.empty() ? journal_path() : std::move(path)) {}

int Journal::add_listener(std::function<void()> cb) {
  int id = next_listener_id_++;
  listeners_.emplace_back(id, std::move(cb));
  return id;
}

void Journal::remove_listener(int id) {
  listeners_.erase(std::remove_if(listeners_.begin(), listeners_.end(),
                                  [id](const auto& l) { return l.first == id; }),
                   listeners_.end());
}

void Journal::remove_listeners() {
  listeners_.clear();
}

void Journal::append(JournalEntry entry) {
  entry.command = clip(entry.command, kCommandMax);
  entry.error = clip(entry.error, kErrorMax);
  if (entry.id.empty()) {
    entry.id = new_uuid();
  }
  std::string line = entry_to_json(entry).dump();
  {
    std::lock_guard lock(mutex_);
    std::filesystem::create_directories(path_.parent_path());
    std::ofstream out(path_, std::ios::binary | std::ios::app);
    out << line << "\n";
    try {
      auto text = read_text_file(path_);
      std::vector<std::string> lines;
      std::istringstream ss(text);
      std::string row;
      while (std::getline(ss, row)) {
        if (!trim(row).empty()) lines.push_back(row);
      }
      if (static_cast<int>(lines.size()) > max_entries + kTrimSlack) {
        std::ostringstream kept;
        int start = static_cast<int>(lines.size()) - max_entries;
        for (int i = start; i < static_cast<int>(lines.size()); ++i) {
          kept << lines[static_cast<std::size_t>(i)] << "\n";
        }
        atomic_write_text(path_, kept.str());
      }
    } catch (...) {
    }
  }
  for (auto& [id, cb] : listeners_) {
    (void)id;
    try {
      cb();
    } catch (...) {
    }
  }
}

std::vector<JournalEntry> Journal::load(int limit) const {
  std::lock_guard lock(mutex_);
  std::vector<JournalEntry> entries;
  if (!std::filesystem::exists(path_)) {
    return entries;
  }
  try {
    auto text = read_text_file(path_);
    std::istringstream ss(text);
    std::string line;
    while (std::getline(ss, line)) {
      line = trim(line);
      if (line.empty()) continue;
      try {
        auto parsed = json::parse(line);
        if (parsed.is_object()) {
          entries.push_back(parse_entry(parsed));
        }
      } catch (...) {
      }
    }
  } catch (...) {
    return {};
  }
  if (limit > 0 && static_cast<int>(entries.size()) > limit) {
    entries.erase(entries.begin(), entries.end() - limit);
  }
  std::reverse(entries.begin(), entries.end());
  return entries;
}

std::map<std::string, JournalEntry> Journal::latest_by_command_id() const {
  std::map<std::string, JournalEntry> latest;
  for (const auto& item : load()) {
    auto cid = trim(item.command_id);
    if (!cid.empty() && !latest.count(cid)) {
      latest.emplace(cid, item);
    }
  }
  return latest;
}

bool Journal::remove(const std::string& id) {
  auto target = trim(id);
  if (target.empty()) return false;
  bool removed = false;
  {
    std::lock_guard lock(mutex_);
    if (!std::filesystem::exists(path_)) return false;
    std::vector<std::string> kept;
    try {
      auto text = read_text_file(path_);
      std::istringstream ss(text);
      std::string line;
      while (std::getline(ss, line)) {
        auto row = trim(line);
        if (row.empty()) continue;
        bool drop = false;
        try {
          auto parsed = json::parse(row);
          if (parsed.is_object() && parsed.value("id", "") == target) {
            drop = true;
          }
        } catch (...) {
        }
        if (drop) {
          removed = true;
          continue;
        }
        kept.push_back(std::move(row));
      }
    } catch (...) {
      return false;
    }
    if (!removed) return false;
    std::ostringstream out;
    for (const auto& row : kept) {
      out << row << "\n";
    }
    atomic_write_text(path_, out.str());
  }
  for (auto& [id, cb] : listeners_) {
    (void)id;
    try {
      cb();
    } catch (...) {
    }
  }
  return true;
}

void Journal::clear() {
  {
    std::lock_guard lock(mutex_);
    atomic_write_text(path_, "");
  }
  for (auto& [id, cb] : listeners_) {
    (void)id;
    try {
      cb();
    } catch (...) {
    }
  }
}

std::string Journal::export_text(const std::vector<JournalEntry>* entries) const {
  auto data = entries ? *entries : load();
    std::ostringstream ss;
    const char* sep = "\n\n────────────────────────────────────────────────\n\n";
    for (std::size_t i = 0; i < data.size(); ++i) {
      if (i) ss << sep;
      ss << data[i].as_text();
    }
  return ss.str();
}

}  // namespace fatty
